jmclient/jmclient/cryptoengine.py

```diff
@@ -283,8 +289,108 @@ def sign_transaction(cls, tx, index, privkey, amount,
         return btc.sign(btc.serialize(tx), index, privkey,
                         hashcode=hashcode, amount=amount, native=True)
 
+class BTC_Timelocked_P2WSH(BTCEngine):
+
+    """
+    In this class many instances of "privkey" or "pubkey" are actually tuples
+    of (privkey, timelock) or (pubkey, timelock)
+    """
+
+    @classproperty
+    def VBYTE(cls):
+        return 100 #bech32 address vbyte
+
+    @classmethod
+    def privkey_to_script(cls, privkey_locktime):
+        privkey, locktime = privkey_locktime
+        pub = cls.privkey_to_pubkey(privkey)
+        return cls.pubkey_to_script((pub, locktime))
+
+    @classmethod
+    def pubkey_to_script(cls, pubkey_locktime):
+        redeem_script = cls.pubkey_to_script_code(pubkey_locktime)
+        return btc.redeem_script_to_p2wsh_script(redeem_script)
+
+    @classmethod
+    def pubkey_to_script_code(cls, pubkey_locktime):
+        pubkey, locktime = pubkey_locktime
+        return btc.mk_freeze_script(pubkey, locktime)
+
+    @classmethod
+    def privkey_to_wif(cls, privkey_locktime):
+        priv, locktime = privkey_locktime
+        return btc.bin_to_b58check(priv, cls.WIF_PREFIX)
+
+    @classmethod
+    def sign_transaction(cls, tx, index, privkey_locktime, amount,
+                         hashcode=btc.SIGHASH_ALL, **kwargs):
+        assert amount is not None
+
+        privkey, locktime = privkey_locktime
+        privkey = hexlify(privkey).decode()
+        pubkey = btc.privkey_to_pubkey(privkey)
+        redeem_script = cls.pubkey_to_script_code((pubkey, locktime))
+        tx = btc.serialize(tx)
+        sig = btc.get_p2sh_signature(tx, index, redeem_script, privkey,
+            amount)
+        return btc.apply_freeze_signature(tx, index, redeem_script, sig)
+
+class BTC_Watchonly_Timelocked_P2WSH(BTC_Timelocked_P2WSH):
+
+    @classmethod
+    def get_watchonly_path(cls, path):
+        #given path is something like "m/49'/1'/0'/0/0"
+        #but watchonly wallet already stores the xpub for "m/49'/1'/0'/"
+        #so to make this work we must chop off the first 3 elements
+        return path[3:]
+
+    @classmethod
+    def derive_bip32_privkey(cls, master_key, path):
+        assert len(path) > 1
+        return cls._walk_bip32_path(master_key, cls.get_watchonly_path(
+            path))[-1]
+
+    @classmethod
+    def privkey_to_script(cls, pubkey_locktime):
+        pub, locktime = pubkey_locktime
+        return cls.pubkey_to_script((pub, locktime))
+
+    @classmethod
+    def privkey_to_wif(cls, privkey_locktime):
```

Comment:
Does this method exist to override the one in `BTCEngine`? I don't actually know the rule about inheritance with classmethods, I guess inheritance works the same as with instance methods? Then it makes sense to me, you want to override this and nullify it. But wouldn't it make more sense to raise an AssertionError than return a null string, since this should never be called, or is there something else to it?

Reply:
I think its called when `wallet-tool.py` wants to print out private keys, and I thought it made sense to not print anything rather than just error.

Reply:
I think calling functions should change their behaviour, recognizing that this is a watch-only wallet, rather than have a function that does return anything privkey-related in this class, even if what is returned is `""` or `"ERROR-STRING-BIG-WARNING"`. Imagine someone writing code that prints out privkeys somewhere (into a temporary file or any kind of weird nonsense) and then imports them somewhere else to generate an address ...
OK, since this is a function creating a WIF, not just a raw key, that scenario doesn't work but I just think it's better to reject any attempt to interact with a private key *as violently as possible*.